examples/src/main/java/com/datastax/oss/driver/examples/concurrent/LimitConcurrencyCustom.java

```diff
@@ -0,0 +1,149 @@
+ *   <li>inserts a TOTAL_NUMBER_OF_INSERTS of rows into the table.
+ * </ul>
+ *
+ * @see <a href="http://datastax.github.io/java-driver/manual/">Java driver online manual</a>
+ */
+public class LimitConcurrencyCustom {
+  private static final int CONCURRENCY_LEVEL = 32;
+  private static final int TOTAL_NUMBER_OF_INSERTS = 10_000;
+  private static final int IN_FLIGHT_REQUESTS = 500;
+  // Semaphore for limiting number of in-flight requests.
+  private static final Semaphore SEMAPHORE = new Semaphore(IN_FLIGHT_REQUESTS);
+
+  // Create CountDownLatch that wait for completion of all pending requests
+  private static final CountDownLatch REQUEST_LATCH = new CountDownLatch(TOTAL_NUMBER_OF_INSERTS);
+
+  public static void main(String[] args) throws InterruptedException {
+
+    try (CqlSession session = new CqlSessionBuilder().build()) {
+      createSchema(session);
+      insertConcurrent(session);
+    }
+  }
+
+  private static void insertConcurrent(CqlSession session) throws InterruptedException {
+    PreparedStatement pst =
+        session.prepare(
+            insertInto("examples", "tbl_sample_kv")
+                .value("id", bindMarker("id"))
+                .value("value", bindMarker("value"))
+                .build());
+
+    // Used to track number of total inserts
+    AtomicInteger insertsCounter = new AtomicInteger();
+    // Used to track threads that were involved in a processing
+    Set<String> threads = new ConcurrentSkipListSet<>();
+
+    // Executor service with CONCURRENCY_LEVEL number of threads that states an upper limit
+    // on number of request in progress.
+    ExecutorService executor = Executors.newFixedThreadPool(CONCURRENCY_LEVEL);
+
+    // For every i we will insert a record to db
+    for (int i = 0; i < TOTAL_NUMBER_OF_INSERTS; i++) {
+      // Before submitting a request, we need to acquire 1 permit.
+      // If there is no permits available it blocks caller thread.
+      SEMAPHORE.acquire();
+      // Copy to final variable for usage in a separate thread
+      final int counter = i;
+
+      // We are running CqlSession.executeAsync in a separate thread pool (executor)
+      CompletableFuture.supplyAsync(
+          () -> {
+            insertsCounter.incrementAndGet();
+            threads.add(Thread.currentThread().getName());
+            CompletableFuture<? extends AsyncResultSet> completableFuture =
+                session
+                    .executeAsync(
+                        pst.bind().setUuid("id", UUID.randomUUID()).setInt("value", counter))
+                    .toCompletableFuture();
+            // Block the current Thread until the result is ready.
+            // When the completableFuture finishes it means that
+            // this thread is free and can pick up another call to CqlSession.executeAsync()
+            AsyncResultSet executedRequest =
+                CompletableFutures.getUninterruptibly(completableFuture);
```

Comment:
I think we should wrap this into a try/finally, to always decrement the counter and release the semaphore.